Improve Null-Like behaviour in comparisons

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This changes the operator binders for `-lt`/`-gt` (etc.) to recognise null-like values the same as it does for `null`, so that it treats them as a `0` value when comparing to a numeric value.

```powershell
[nullstring]::Value -gt 3 # False
[dbnull]::Value -gt 2 # False
```

## PR Context

Fixes #10404

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:sparkles: Handle null-like as null for -gt/-lt
Handling null-like items (DBNull.Value/NullString.Value) the same way
as true null here allows database-interfacing code to do
numeric comparisons such as `$dbEntry.Value -gt 4` and get
a sensible response of true/false depending on operator.
Without this change, these operations would result in errors.

vexx32 committed:
:white_check_mark: Add tests

Use -ge not -eq for test
For fairly plain reasons, PS doesn't consider $null as equal to zero
This is only a good approximation in definitively numeric circumstances

PoshChan commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3170,18 +3170,23 @@ private static Expression CompareWithZero(DynamicMetaObject target, Func<Express
             return comparer(target.Expression.Cast(target.LimitType), ExpressionCache.Constant(0).Cast(target.LimitType)).Cast(typeof(object));
```
@vexx32, your last commit had 12 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ $null } should be greater than -1
```powershell
Cannot compare " $null " because it is not IComparable.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 115
115:             $Value -gt -1 | Should -BeTrue

```

PoshChan replied:
@vexx32, your last commit had 12 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ $null } should be greater than -1
```powershell
Cannot compare " $null " because it is not IComparable.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 115
115:             $Value -gt -1 | Should -BeTrue

```

PoshChan replied:
@vexx32, your last commit had 12 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ $null } should be greater than -1
```powershell
Cannot compare " $null " because it is not IComparable.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 115
115:             $Value -gt -1 | Should -BeTrue

```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-macos`
Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-windows`
Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-linux`
Null Representatives.Numeric Comparisons.{ $null } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             0 -eq $Value.InvokeReturnAsIs() | Should -BeTrue
```

iSazonov commented:
One question is performance degradation.
I guess we need replace
`public static bool IsNullLike(object obj) => obj == DBNull.Value || obj == NullString.Value || IsNull(obj);` 
with
`public static bool IsNullLike(object obj) => IsNull(obj) || obj == DBNull.Value || obj == NullString.Value;`

vexx32 committed:
:recycle: Refactor IsNullLike for fast path

vexx32 commented:
Good call, @iSazonov, made that change. 💖

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3170,18 +3170,23 @@ private static Expression CompareWithZero(DynamicMetaObject target, Func<Express
+                    ? CompareWithZero(target, Expression.LessThan)
+                    : arg.LimitType.IsNumeric()
+                        ? CompareWithZero(arg, Expression.GreaterThanOrEqual)
+                        : !LanguagePrimitives.IsNullLike(arg.Value)
```
Can you please reverse this ternary expression? Instead of `!<....> ? a : b`, make it `<> ? b : a`. It's easier to read the code that way.

vexx32 replied:
Done!

daxian-dbw commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -3190,18 +3195,23 @@ private static Expression CompareWithZero(DynamicMetaObject target, Func<Express
+                    ? CompareWithZero(target, Expression.LessThan)
+                    : arg.LimitType.IsNumeric()
+                        ? CompareWithZero(arg, Expression.GreaterThanOrEqual)
+                        : !LanguagePrimitives.IsNullLike(target.Value)
```
Same here and the rest places.

vexx32 replied:
and done! 🙂

vexx32 committed:
:recycle: Invert confusing ternary expressions

daxian-dbw commented:
I played the changes a bit and found an issue:

```powershell
> [System.DBNull]::Value -ge [NullString]::Value
True
> [NullString]::Value -le [System.DBNull]::Value
True
```
However, we don't consider `[DBNull]::Value -eq [NullString]::Value` , see https://github.com/PowerShell/PowerShell/pull/9794/files#diff-c6cad5a3db5d6bcb21228e483b7506d1R31-R42

vexx32 commented:
@daxian-dbw yeah I can see how that would arise, for sure. It is an anomaly, but it makes some sense here.

With `-ge` operators we are explicitly using either only a string or a numeric comparison mode. As a result, both operands must be converted into a numeric form or a string form for the comparison.

Is there a better option available for us to use here?

iSazonov commented:
Taking that `[DBNull]::Value -eq [NullString]::Value` is false we should fix `-ge` and `-le` I believe.

vexx32 commented:
Is there a sensible result for either comparison? In what sense is one version of a null value numerically greater or lesser than another? Or should we be doing something very unusual and always returning $false for any such comparison?

In other words, should this be the case?
```ps
$NullString = [NullString]::Value
$DBNull = [DBNull]::Value

$NullString -ge $DBNull # -> false
$NullString -le $DBNull # -> false
$NullString -gt $DBNull # -> false
$NullString -lt $DBNull # -> false
$DBNull -ge $NullString # -> false
$DBNull -le $NullString # -> false
$DBNull -gt $NullString # -> false
$DBNull -lt $NullString # -> false
```

While this is possible, I'm not sure it's sensible. `-ge`, `-gt`, and their counterparts are deliberately implemented from the beginning as different from `-eq` and `-ne`. The latter convert types when needed only; the former will always evaluate as either a numeric or a string comparison.

iSazonov commented:
We could add samples for $null in the list.

vexx32 commented:
@iSazonov all comparisons with null-like values and $null are $true. I'll have to verify -ge and -le but I believe they evaluate as equal in all cases. :slightly_smiling_face:

vexx32 commented:
@iSazonov updated list of what the suggested change `$DBNull -ge $NullString => $false `_and_ `$DBNull -le $nullString=> $false` would look like with $null included:

(Personally, I don't think this makes sense; these comparisons have always worked on a slightly different basis to `-eq`/`-ne` as they're stringly numeric/string comparisons anyway.)

```ps
$NullString = [NullString]::Value
$DBNull = [DBNull]::Value

$NullString -ge $DBNull # -> false
$NullString -le $DBNull # -> false
$NullString -ge $null # -> true
$NullString -le $null # -> true

$NullString -gt $DBNull # -> false
$NullString -lt $DBNull # -> false
$NullString -gt $null # -> true (?)
$NullString -lt $null # -> false (?)

$DBNull -ge $NullString # -> false
$DBNull -le $NullString # -> false
$DBNull -ge $null # -> true
$DBNull -le $null # -> true

$DBNull -gt $NullString # -> false
$DBNull -lt $NullString # -> false
$DBNull -gt $null # -> true (?)
$DBNull -lt $null # -> false (?)
```

I think that _despite_ $NullString and $DBNull being not equal to each other, comparisons with `-ge` or `-le` (or indeed `-gt` or `-lt`) are explicitly either based on numeric value or string sort order. As such, they should convert to `0` or empty string and give a `$true` result for both `-ge`/`-le` and a `$false` result for `-gt/-lt` — this is how the current implementation in this branch would work.

Thoughts, @daxian-dbw? 🙂

vexx32 commented:
@daxian-dbw @SteveL-MSFT friendly ping for review! 💖 😊

SteveL-MSFT commented on test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1:
```diff
@@ -86,4 +86,33 @@ Describe 'Null Representatives' -Tags 'CI' {
+        It '<Value> should be considered greater than or equal to 0' -TestCases $TestValues {
+            param($Value)
+
+            0 -ge $Value.InvokeReturnAsIs() | Should -BeTrue
```
Any particular reason `0` is left hand side unlike the tests below?  For completeness, it seems `-le` should be covered.

vexx32 replied:
Not that I can recall. Will add the `-le` test as well. 🙂

vexx32 replied:
Addressed and pushed new commits. Also rebased since this branch is pretty old. 🙂

vexx32 committed:
:white_check_mark: Update tests
Address review comments, add test for -le as well
Swap LHS<>RHS for -le / -ge tests, null-ish thing on LHS

PoshChan commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -1071,7 +1071,7 @@ internal static bool IsTrue(IList objectArray)
         /// </summary>
```
@vexx32, your last commit had 4 failures in `PowerShell-CI-macos`
Null Representatives.Numeric Comparisons.{ $null } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-linux`
Null Representatives.Numeric Comparisons.{ $null } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-macos`
Null Representatives.Numeric Comparisons.{ $null } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-linux`
Null Representatives.Numeric Comparisons.{ $null } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [DBNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [NullString]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```
Null Representatives.Numeric Comparisons.{ [AutomationNull]::Value } should be considered greater than or equal to 0
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NullRepresentatives.Tests.ps1: line 103
103:             $Value.InvokeReturnAsIs() -ge 0 | Should -BeTrue
```

vexx32 commented:
@PoshChan please retry macos, linux

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-macOS, PowerShell-CI-Linux`

vexx32 committed:
:bug: Fix issue with binders
Binders were returning the wrong result in some cases
When `arg` was numeric the test was creating a LessThan/GreaterThan
expression instead of the Less/GreaterThanOrEqual
that would be accurate.

vexx32 commented:
Okay, finally got the binder issue sorted. Very odd that it wasn't surfacing on the Windows VM; it was showing locally on my Windows machine when I checked.

The logic paths there are horribly confusing, and that nested tree of ternary needs a lot of puzzling out. I'm sorely tempted to say we need to refactor it completely so that we can actually see what's going on there, but I don't really see much of a better way to write it unfortunately.

iSazonov commented:
@daxian-dbw Please look the last binder fix. It is very sensitive code.

vexx32 commented:
@daxian-dbw @iSazonov I'll note this in the PR description too, but that last fix resolves quite a long-standing issue with `-ge` comparisons with `$null`.

![image](https://user-images.githubusercontent.com/32407840/66799016-8484f580-eedd-11e9-9c1d-540271075e54.png)

(This is in pwsh v7-preview3.)

With the most recent commit this is no longer an issue, and is also resolved for the above case as well.

iSazonov commented:
/cc @mklement0 You might be interested in last commit.

iSazonov commented:
@vexx32 You could add a test for your last comment - to explicitly demo the side effect.

vexx32 committed:
:white_check_mark: Add mirror tests for comparison

vexx32 commented:
@iSazonov that should do the trick. If you want me to break them out completely separately, let me know, but I figured I may as well make the most of the testcases setup I have going already. 🙂

mklement0 commented:
@vexx32  The problem with making `$null -eq 0` evaluate to `$true` is that it breaks existing code that uses `$null -eq ...` to test the RHS for being truly `$null`.

Aside from potentially breaking existing code, this would leave PowerShell users with no straightforward way to test a value for `$null`.

Generally, it is the LHS that determines the type for the comparison,  so from that perspective `$null -eq 0` evaluating to `$false` makes sense (only the unequivocally _arithmetic_ operators `/` and `-` seems to coerce the LHS to a number too).

Even if `$null -eq 0` remains `$false`, your PR will result in  the inability to test for `$null` specifically, as distinct from other null-like values such as `[DbNull]::Value` - though I suppose that's unavoidable, at least when testing with `-eq`

@KirkMunro's PR https://github.com/PowerShell/PowerShell/pull/10704 comes to the rescue here, which will allow testing with `... -is $null`.

An obscure workaround is to use `[object]::ReferenceEquals()`.

iSazonov commented:
It looks like huge breaking change.

mklement0 commented:
Agreed.

To put it differently: The existing asymmetry of `-eq`'s behavior - as surprising as it may be - is a feature, not a bug.

vexx32 commented:
@mklement0 I have not modified the `-eq` / `-ne` binders here. They are unaffected. The asymmetry only pertains to `-ge` / `-le`

This is why I have deemed the asymmetry worth changing. As you say, in an **arithmetic** case, the LHS is converted to a number regardless, and this has always been the case with these operators is the past when given a null value. However, the result does not match with all other conversions from null, and instead treats null as _less than_ zero, despite the fact that null values are unequivocally treated as greater than all negative values already.

vexx32 commented:
To put it another way.... The expression `0 -eq $null` is `$true`

Yet (currently) `$null -ge 0` is `$false`

... That's what's wrong there.

iSazonov commented:
> ... That's what's wrong there.

I'd ask what was wrong in scripts that we should fix? :-) It feels like we're trying to improve something that doesn't need improvement :-)

vexx32 commented:
Perhaps. I'd argue that getting mathematical consistency should be a a desirable target. @daxian-dbw , your thoughts?

KirkMunro commented:
I'd argue that like all comparison operators in PowerShell, the type used for the comparison, the type used in the comparison is consistently driven by the type of the operand on the left-hand side.

With that, I expect `$null -ge 0` to always return `$false`, and `0 -le $null` to always return true. From a mathematical standpoint that shouldn't confuse you once you learn that the type used in the comparison is driven by the operand on the left-hand side. That's just a programming rule you need to understand, not a gotcha or a trick.

vexx32 commented:
Hmm, that is usually true. But if it were, `0 -ge $null` should _also_ be true. And it's not.

mklement0 commented:
> The asymmetry only pertains to -ge / -le

I see, thanks; I hadn't noticed that distinction.

That's less problematic, but to me that still amounts to a problematic inconsistency - though I just realized we already have that inconsistency with `-le`.

If I understand you correctly, your change would then result in the following:

```powershell
$null -eq 0 # -> $false

$null -ge 0 # -> $true - this is what your change will bring

$null -le 0 # -> $true  # !! This is already the case, which I didn't expect.
$null -lt 0 # -> $true  # !! This is currently the case, which I didn't expect.
```

If it is the quasi-zero-ness of `$null` that makes the `-le` and `-ge` cases return `$true`, then why wouldn't the `-eq` case return `$true` also?

In the abstract,  it would make more sense to let the type (non-type) of the LHS - `$null` - in this case consistently drive the behavior, so that the `-le` and `-ge` cases _fail_ with `Cannot compare $null because it is not IComparable`.

Note that `-ge` and `-le` don't  _just_ arithmetic meaning, and given that `$null` is the _LHS_ you cannot, strictly speaking, make that assumption (the way it is made with the non-overloaded `/` and `-` operators).

For instance, `'10' -ge 2` is `$false`, because _lexical_ comparison is performed - the LHS type drives the interpretation of `-ge`.

In short: Whether `-ge` / `-le` / `-gt` / `-lt` can be meaningfully applied depends on whether the type at hand implements `IComparable`.

----

I see the following ways to resolve this.

* Do nothing - live with and document the inconsistency (not my preference).

* Consistently make the LHS type/non-type dictate the behavior, which means switching the `-ge` / `-le` / `-gt` / `-lt` cases to _failing_ with `Cannot compare $null because it is not IComparable` 

  *  I can see how there's code out there that does `if ($uninitializedVariable -le 0) ...` (though the `if ($uninitializedVariable -ge 0) ...` currently doesn't work meaningfully - which is what motivates your fix, and neither does `if ($uninitializedVariable -lt 0) ...`).

* Leave the `$null -eq 0` -> `$false` exception in place to preserve backward compatibility and make only the `-ge` / `-le` / `-gt` / `-lt` cases behave _numerically_, meaning that `$null` as the LHS is treated like `0` - that amounts to your fix, @vexx32, _plus_ also making `$null -lt 0` return `$false`.

vexx32 commented:
The `-eq` case doesn't return true because it's not explicitly a mathematical operation, equality is more than just math. A greater/less than comparison, though, is a mathematical operation. Or string sorting, in our cases.

Pretty sure I have your last option there completely implemented. There should be a test confirming your point about `$null -lt 0` but I will double check.

mklement0 commented:
Good point about `0 -ge $null` / `0 -gt $null` / `0 -le $null` currently misbehaving too, @vexx32

There, the case is clear-cut: the numeric LHS should coerce the `$null` to a number, which should be `0`.

Would this need to be a separate fix?

> A greater/less than comparison, though, is a mathematical operation.

Well, it's whatever the type's `IComparable` implementation defines it to be - and that doesn't have to be _numeric_ comparison, which is what we're - inconsistently - assuming here.

> Or string sorting, in our cases.

Again, this comes from the _LHS_ dictating the behavior - and the consistent implementation of that for `$null` would be to say: you cannot compare `$null` (except to itself, with `-eq`).

vexx32 commented:
I am happy to commit that fix separately is another PR if we prefer. Now that I see where it went wrong, it's a fairly simple asymmetry to resolve.

iSazonov commented:
> Would this need to be a separate fix?

Will it again a breaking change?

mklement0 commented:
@iSazonov: 

It would technically be a breaking changing, but given that the current behavior is useless in the cases we'd be fixing, I don't expect that to be a problem.

Let me try to summarize the discussion, to make sure we're on the same page:

* `$null -eq 0` retains its current behavior of evaluating to `$false`, because `$null` as the LHS with `-eq` must continue to function as the is-it-`$null`[-like] test.

* `-ge` / `-le` / `-gt` / `-lt` behavior will be modified so that `$null` on _either_ side is implicitly treated like `0`, to make the current behavior with these operators consistent. 

* While this distinction between `-eq` vs.  `-ge` / `-le` / `-gt` / `-lt` behavior amounts to an overall inconsistency with respect to letting the LHS drive type coercions, given the current behavior, it strikes me as a reasonable forward:
   * The truly consistent behavior - failing with `Cannot compare $null because it is not IComparable` with `$null` as the LHS when combined with `-ge` / `-le` / `-gt` / `-lt` - is probably not that useful and _may_ break existing scripts (though the treat-as-`0` behavior currently only works in _some_ cases - see tests below).
   * Absent that, I suppose another option is to _always_ return `$false` if the RHS isn't null-like, but that could again break existing scripts. 

At the very least, the cases where `$null`[-like] is the _RHS_ and currently doesn't behave like `0` with a numeric LHS should be fixed.

To put it in terms of Pester tests:

**Null-like as the _RHS_**:

```powershell
# These currently fail and should be fixed in any event.
# $null[-like] should consistently be coerced to 0 in these cases.
0 -gt $null | Should -BeFalse
0 -eq $null | Should -BeTrue
0 -gt $null | Should -BeFalse
0 -le $null | Should -BeTrue

# These already work as expected.
0 -ge $null | Should -BeTrue
0 -lt $null | Should -BeFalse
```

**Null-like as the _LHS_**:

```powershell
# Passes, and shouldn't change.
# Anything other than a null-like type on the RHS should result in $false
$null -eq 0 | Should -BeFalse

# These are currently the only cases with $null as the LHS were
# numeric interpretation works as expected:
$null -le 0 | Should -BeTrue
$null -lt 0 | Should -BeFalse

# These currently fail from a treat-null-as-0 perspective.
$null -ge 0 | Should -BeTrue
$null -lt 0 | Should -BeFalse
```

iSazonov commented:
@mklement0 Thanks for the summarization!
This is twice in example above
> 0 -gt $null | Should -BeFalse

Difference `$null -eq 0` and `0 -eq $null` is too unexpected and I believe this should not be in the scripting language, it is very bad UX.

mklement0 commented:
Thanks, @iSazonov, I've removed the duplicate test from my previous comment.

> Difference $null -eq 0 and 0 -eq $null is too unexpected

I agree that it is unfortunate, but not _fundamentally_ - asymmetrical `-eq` behavior is very much part of PowerShell (see below) - but because it creates a surprising behavioral difference between `-eq` / `-ne` on the one hand and `-ge` / `-le` / `-gt` / `-lt` on the other, which doesn't exist for other types (again, if `$null -ge 0` is `$true` because `$null` is treated like `0`, why wouldn't `$null -eq 0` be `$true` as well?)

However, we cannot change `$null -eq 0` / `$null -ne 0`, because historically `$null -eq ...` has been the only robust test for comparing any value to `$null`.

Pragmatically speaking, interpretation of LHS `$null` as `0` with `-ge` / `-le` / `-gt` / `-lt` does make sense, especially if you consider use of uninitialized variables that evaluate to `$null` on the LHS (e.g., `if ($uninitializedVar -gt 0)`), but the - unavoidable - downside is the different behavior of  `$uninitializedVar -eq ...` / `$uninitializedVar -ne ...`

As for the fundamental asymmetry with `-eq`, due to type coercions: consider the following case:
```powershell
PS> 2 -eq '+2'
True

PS> '+2' -eq 2
False
```

vexx32 commented:
Yeah, we can't avoid that difference. If we try to mess around with that we'll get in the way of PS's array-filtering semantics with `-eq` and similar operators, as well as the type coercion semantics.

It is also a bit confusing, but it does follow clearly defined rules -- unlike the current behaviour with `-ge`/`-le` 😁

iSazonov commented:
> Yeah, we can't avoid that difference.

We have been living with this for a long time.
I think the best way for the project with a long history is to correct specific problems that arise in practical scenarios. As long as there is no significant business scenario, it makes no sense to make a breaking change. I would prefer such a conservative way until significant improvement ideas appear for which users will be ready to make significant changes in their bussines.

vexx32 commented:
I understand your concerns there @iSazonov. But I would prefer the language behave as consistently as possible; the fewer strange exceptions to the rules are, the better.

@daxian-dbw can you or the @PowerShellTeam weigh in on this aspect? It's not critical to this PR, but I feel it would make sense to take care of within the scope of this PR if we're going to do so.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and is concerned that changing this behavior will break other scenarios that rely on the variants of null to be different due to loss of information.  We did not conclude on this but agreed it isn't necessary for PS7.

vexx32 commented:
@SteveL-MSFT is that specifically about having `-ge` and `-le` treat $null as `0` that we're wanting to retain? Or are you talking about this PR as a whole? (If so; the current behaviour for NullString/DBNull with `-ge` and `-le` is simply to throw an error, so I don't think we'd be breaking much here.) 🙂

SteveL-MSFT commented:
cc @JamesWTruher

SteveL-MSFT commented:
@PowerShell/powershell-committee has discussed this last week as part of a broader `[DBNull]` discussion.  Concern is whether we made the right decision to allow `[DBNull]::value -eq $null` which has brought up many other related issues and also introducing regressions.  We agreed to revert the change where `[DBNull]::value -eq $null` and encourage continued discussion particularly with DB users of PowerShell to understand the pain points related to use of DBNull and how to best address it in PowerShell thinking holistically across the language including other operators and method invocations, etc... that have come up.

iSazonov commented:
> encourage continued discussion particularly with DB users of PowerShell

Please add reference to the issue.

SteveL-MSFT commented:
Created https://github.com/PowerShell/PowerShell/issues/11604
